File: framework/wcf/Extensibility/Transactions/TransactionMessagePropertyUdpTransport/CS/UdpTransport/SampleProfileUdpBindingConfigurationElement.cs
//  Copyright (c) Microsoft Corporation. All rights reserved.

using System;
using System.Configuration;
using System.Globalization;
using System.ServiceModel.Channels;
using System.ServiceModel.Configuration;

namespace Microsoft.ServiceModel.Samples
{
    public class SampleProfileUdpBindingConfigurationElement : StandardBindingElement
    {
        public SampleProfileUdpBindingConfigurationElement(string configurationName)
            : base(configurationName)
        {
        }

        public SampleProfileUdpBindingConfigurationElement()
            : this(null)
        {
        }

        protected override Type BindingElementType
        {
            get { return typeof(SampleProfileUdpBinding); }
        }

        [ConfigurationProperty(UdpConfigurationStrings.OrderedSession, DefaultValue = UdpDefaults.OrderedSession)]
        public bool OrderedSession
        {
            get { return (bool)base[UdpConfigurationStrings.OrderedSession]; }
            set { base[UdpConfigurationStrings.OrderedSession] = value; }
        }

        [ConfigurationProperty(UdpConfigurationStrings.ReliableSessionEnabled, DefaultValue = UdpDefaults.ReliableSessionEnabled)]
        public bool ReliableSessionEnabled
        {
            get { return (bool)base[UdpConfigurationStrings.ReliableSessionEnabled]; }
            set { base[UdpConfigurationStrings.ReliableSessionEnabled] = value; }
        }

        [ConfigurationProperty(UdpConfigurationStrings.SessionInactivityTimeout, DefaultValue = UdpDefaults.SessionInactivityTimeoutString)]
        [TimeSpanValidator(MinValueString = "00:00:00")]
        public TimeSpan SessionInactivityTimeout
        {
            get { return (TimeSpan)base[UdpConfigurationStrings.SessionInactivityTimeout]; }
            set { base[UdpConfigurationStrings.SessionInactivityTimeout] = value; }
        }

        [ConfigurationProperty(UdpConfigurationStrings.ClientBaseAddress, DefaultValue = null)]
        public Uri ClientBaseAddress
        {
            get { return (Uri)base[UdpConfigurationStrings.ClientBaseAddress]; }
            set { base[UdpConfigurationStrings.ClientBaseAddress] = value; }
        }

        protected override ConfigurationPropertyCollection Properties
        {
            get
            {
                ConfigurationPropertyCollection properties = base.Properties;
                properties.Add(new ConfigurationProperty(UdpConfigurationStrings.OrderedSession, 
                    typeof(Boolean), UdpDefaults.OrderedSession, null, null, ConfigurationPropertyOptions.None));
                properties.Add(new ConfigurationProperty(UdpConfigurationStrings.ReliableSessionEnabled, 
                    typeof(Boolean), UdpDefaults.ReliableSessionEnabled, null, null, ConfigurationPropertyOptions.None));
                properties.Add(new ConfigurationProperty(UdpConfigurationStrings.SessionInactivityTimeout, 
                    typeof(TimeSpan), TimeSpan.Parse(UdpDefaults.SessionInactivityTimeoutString), null, 
                    new TimeSpanValidator(TimeSpan.Parse("00:00:00"), TimeSpan.Parse("10675199.02:48:05.4775807"), false), 
                    ConfigurationPropertyOptions.None));
                properties.Add(new ConfigurationProperty(UdpConfigurationStrings.ClientBaseAddress,
                    typeof(System.Uri), null, null, null, ConfigurationPropertyOptions.None));

                return properties;
            }
        }        
        
        protected override void InitializeFrom(Binding binding)
        {
            base.InitializeFrom(binding);
            SampleProfileUdpBinding udpBinding = (SampleProfileUdpBinding)binding;

            this.OrderedSession = udpBinding.OrderedSession;
            this.ReliableSessionEnabled = udpBinding.ReliableSessionEnabled;
            this.SessionInactivityTimeout = udpBinding.SessionInactivityTimeout;
            if (udpBinding.ClientBaseAddress != null)
                this.ClientBaseAddress = udpBinding.ClientBaseAddress;
        }

        protected override void OnApplyConfiguration(Binding binding)
        {
            if (binding == null)
                throw new ArgumentNullException("binding");

            if (binding.GetType() != typeof(SampleProfileUdpBinding))
            {
                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture,
                    "Invalid type for binding. Expected type: {0}. Type passed in: {1}.",
                    typeof(SampleProfileUdpBinding).AssemblyQualifiedName,
                    binding.GetType().AssemblyQualifiedName));
            }
            SampleProfileUdpBinding udpBinding = (SampleProfileUdpBinding)binding;

            udpBinding.OrderedSession = this.OrderedSession;
            udpBinding.ReliableSessionEnabled = this.ReliableSessionEnabled;
            udpBinding.SessionInactivityTimeout = this.SessionInactivityTimeout;
            if (this.ClientBaseAddress != null)
                udpBinding.ClientBaseAddress = ClientBaseAddress;
        }
    }
}

File: framework/wcf/Extensibility/Transactions/TransactionMessagePropertyUdpTransport/VB/UdpTransport/SampleProfileUdpBindingConfigurationElement.vb
'  Copyright (c) Microsoft Corporation. All rights reserved.

Imports System.Configuration
Imports System.Globalization
Imports System.ServiceModel.Channels
Imports System.ServiceModel.Configuration

Namespace Microsoft.ServiceModel.Samples
	Public Class SampleProfileUdpBindingConfigurationElement
		Inherits StandardBindingElement
		Public Sub New(ByVal configurationName As String)
			MyBase.New(configurationName)
		End Sub

		Public Sub New()
			Me.New(Nothing)
		End Sub

		Protected Overrides ReadOnly Property BindingElementType() As Type
			Get
				Return GetType(SampleProfileUdpBinding)
			End Get
		End Property

		<ConfigurationProperty(UdpConfigurationStrings.OrderedSession, DefaultValue := UdpDefaults.OrderedSession)> _
		Public Property OrderedSession() As Boolean
			Get
				Return CBool(MyBase.Item(UdpConfigurationStrings.OrderedSession))
			End Get
			Set(ByVal value As Boolean)
				MyBase.Item(UdpConfigurationStrings.OrderedSession) = value
			End Set
		End Property

        <ConfigurationProperty(UdpConfigurationStrings.ReliableSessionEnabled,
            DefaultValue:=UdpDefaults.ReliableSessionEnabled)> _
        Public Property ReliableSessionEnabled() As Boolean
            Get
                Return CBool(MyBase.Item(UdpConfigurationStrings.ReliableSessionEnabled))
            End Get
            Set(ByVal value As Boolean)
                MyBase.Item(UdpConfigurationStrings.ReliableSessionEnabled) = value
            End Set
        End Property

        <ConfigurationProperty(UdpConfigurationStrings.SessionInactivityTimeout,
            DefaultValue:=UdpDefaults.SessionInactivityTimeoutString),
        TimeSpanValidator(MinValueString:="00:00:00")> _
        Public Property SessionInactivityTimeout() As TimeSpan
            Get
                Return CType(MyBase.Item(UdpConfigurationStrings.SessionInactivityTimeout), TimeSpan)
            End Get
            Set(ByVal value As TimeSpan)
                MyBase.Item(UdpConfigurationStrings.SessionInactivityTimeout) = value
            End Set
        End Property

		<ConfigurationProperty(UdpConfigurationStrings.ClientBaseAddress, DefaultValue := Nothing)> _
		Public Property ClientBaseAddress() As Uri
			Get
				Return CType(MyBase.Item(UdpConfigurationStrings.ClientBaseAddress), Uri)
			End Get
			Set(ByVal value As Uri)
				MyBase.Item(UdpConfigurationStrings.ClientBaseAddress) = value
			End Set
		End Property

		Protected Overrides ReadOnly Property Properties() As ConfigurationPropertyCollection
			Get

                Dim properties_local As ConfigurationPropertyCollection = MyBase.Properties
                With properties_local
                    .Add(New ConfigurationProperty(UdpConfigurationStrings.OrderedSession, GetType(Boolean),
                                                   UdpDefaults.OrderedSession, Nothing, Nothing,
                                                   ConfigurationPropertyOptions.None))
                    .Add(New ConfigurationProperty(UdpConfigurationStrings.ReliableSessionEnabled,
                                                   GetType(Boolean), UdpDefaults.ReliableSessionEnabled,
                                                   Nothing, Nothing, ConfigurationPropertyOptions.None))
                    .Add(New ConfigurationProperty(UdpConfigurationStrings.SessionInactivityTimeout,
                                                   GetType(TimeSpan), TimeSpan.Parse(UdpDefaults.SessionInactivityTimeoutString),
                                                   Nothing, New TimeSpanValidator(TimeSpan.Parse("00:00:00"),
                                                                                  TimeSpan.Parse("10675199.02:48:05.4775807"), False),
                                                                              ConfigurationPropertyOptions.None))
                    .Add(New ConfigurationProperty(UdpConfigurationStrings.ClientBaseAddress, GetType(Uri), Nothing,
                                                   Nothing, Nothing,
                                                   ConfigurationPropertyOptions.None))
                End With
                Return properties_local
            End Get
		End Property

		Protected Overrides Sub InitializeFrom(ByVal binding As Binding)
			MyBase.InitializeFrom(binding)
			Dim udpBinding As SampleProfileUdpBinding = CType(binding, SampleProfileUdpBinding)

			Me.OrderedSession = udpBinding.OrderedSession
			Me.ReliableSessionEnabled = udpBinding.ReliableSessionEnabled
			Me.SessionInactivityTimeout = udpBinding.SessionInactivityTimeout
			If udpBinding.ClientBaseAddress IsNot Nothing Then
				Me.ClientBaseAddress = udpBinding.ClientBaseAddress
			End If
		End Sub

		Protected Overrides Sub OnApplyConfiguration(ByVal binding As Binding)
			If binding Is Nothing Then
				Throw New ArgumentNullException("binding")
			End If

			If binding.GetType() IsNot GetType(SampleProfileUdpBinding) Then
                Throw New ArgumentException(String.Format(CultureInfo.CurrentCulture,
                                                          "Invalid type for binding. Expected type: {0}. Type passed in: {1}.",
                                                          GetType(SampleProfileUdpBinding).AssemblyQualifiedName,
                                                          binding.GetType().AssemblyQualifiedName))
			End If
			Dim udpBinding As SampleProfileUdpBinding = CType(binding, SampleProfileUdpBinding)
            With udpBinding
                .OrderedSession = Me.OrderedSession
                .ReliableSessionEnabled = Me.ReliableSessionEnabled
                .SessionInactivityTimeout = Me.SessionInactivityTimeout
            End With
            If Me.ClientBaseAddress IsNot Nothing Then
                udpBinding.ClientBaseAddress = ClientBaseAddress
            End If
        End Sub
	End Class
End Namespace
